demo/rust/filedemo/src/main.rs

```diff
@@ -0,0 +1,168 @@
+*/
+
+use clap::{App, Arg};
+use itertools::Itertools;
+use picovoice::{rhino::RhinoInference, PicovoiceBuilder};
+use rodio::{source::Source, Decoder};
+use std::fs::File;
+use std::io::BufReader;
+use std::path::PathBuf;
+
+fn picovoice_demo(
+    input_audio_path: PathBuf,
+    keyword_path: &str,
+    context_path: &str,
+    porcupine_model_path: Option<&str>,
+    rhino_model_path: Option<&str>,
+    porcupine_sensitivity: Option<f32>,
+    rhino_sensitivity: Option<f32>,
+) {
+    let soundfile = BufReader::new(File::open(input_audio_path).unwrap());
+    let audiosource = Decoder::new(soundfile).unwrap();
+
+    let wake_word_callback = || println!("[wake word]");
+    let inference_callback = |inference: RhinoInference| {
+        if inference.is_understood {
+            println!("Inferred:");
+            println!("{{");
+            println!("\tintent : '{}'", inference.intent.unwrap());
+            println!("\tslots : {{");
+            for (slot, value) in inference.slots.iter() {
+                println!("\t\t{} : {}", slot, value);
+            }
+            println!("\t}}");
+            println!("}}\n");
+        } else {
+            println!("Did not understand the command");
+        }
+    };
+
+    let mut picovoice_builder = PicovoiceBuilder::new(
+        keyword_path,
+        wake_word_callback,
+        context_path,
+        inference_callback,
+    );
+
+    if let Some(porcupine_model_path) = porcupine_model_path {
+        picovoice_builder = picovoice_builder.porcupine_model_path(porcupine_model_path);
+    }
+    if let Some(rhino_model_path) = rhino_model_path {
+        picovoice_builder = picovoice_builder.rhino_model_path(rhino_model_path);
+    }
+    if let Some(porcupine_sensitivity) = porcupine_sensitivity {
+        picovoice_builder = picovoice_builder.porcupine_sensitivity(porcupine_sensitivity);
+    }
+    if let Some(rhino_sensitivity) = rhino_sensitivity {
+        picovoice_builder = picovoice_builder.rhino_sensitivity(rhino_sensitivity);
+    }
+
+    let mut picovoice = picovoice_builder
+        .init()
+        .expect("Failed to create Picovoice");
+
+    if picovoice.sample_rate() != audiosource.sample_rate() {
+        panic!(
+            "Audio file should have the expected sample rate of {}, got {}",
+            picovoice.sample_rate(),
+            audiosource.sample_rate()
+        );
+    }
+
+    for frame in &audiosource.chunks(picovoice.frame_length() as usize) {
+        let frame = frame.collect_vec();
+        if frame.len() == picovoice.frame_length() as usize {
+            picovoice.process(&frame).unwrap();
+        }
+    }
+}
+
+fn main() {
+    let matches = App::new("Picovoice Rhino Rust File Demo")
+        .arg(
+            Arg::with_name("input_audio_path")
+            .long("input_audio_path")
+            .value_name("PATH")
+            .help("Path to input audio file (mono, WAV, 16-bit, 16kHz)")
```

Comment:
do we enforce the audio format requirements in the demo? I don't know much about the audio library you're using, but if there's a way to let the user know their file isn't valid, that would be ideal

Reply:
Technically the audio reading library used here can handle any standard audio file format (and any sample type wrt. WAV), the only thing we need is the 16kHz sample rate which we do check for

Reply:
but no matter what they input, we get our required format for sure? like if they input a 16kHz mp3 do we report an error or take it?

Reply:
We will take it and it will work!

Reply:
nifty! In that case, can you either update the help to indicate that more formats are accepted, or block them from sending anything but WAV?

Reply:
Actually I take it back, I played around a bit more with the formats just to double check and it seems I just got lucky with the ones I tested before, not everything "just works". I'll just restrict it to what's in the help as suggested.

Reply:
I'll make this same change in the Rhino & Porcupine demos also